src/main/kotlin/com/nikitanazarov/coroutinestacks/CoroutineStacksPanel.kt

```diff
@@ -1,64 +1,386 @@
+        val coroutineStacksWindowHeader = Box.createHorizontalBox()
+        val dispatchers = dispatcherToCoroutineDataList.keys.toTypedArray()
+        val dispatcherDropdownMenu = ComboBox(dispatchers)
+
+        dispatcherDropdownMenu.addActionListener {
+            val selectedDispatcher = dispatcherDropdownMenu.selectedItem as String
+            val coroutineDataList = dispatcherToCoroutineDataList[selectedDispatcher]
+            if (!coroutineDataList.isNullOrEmpty()) {
+                updateCoroutineStackForest(coroutineDataList, suspendContextImpl)
+            }
+        }
+
+        val comboBoxSize = Dimension(Constants.comboBoxHeight, Constants.comboBoxWidth)
+        dispatcherDropdownMenu.apply {
+            preferredSize = comboBoxSize
+            maximumSize = comboBoxSize
+            minimumSize = comboBoxSize
+        }
+        coroutineStacksWindowHeader.add(dispatcherLabel)
+        coroutineStacksWindowHeader.add(dispatcherDropdownMenu)
+
+        dispatcherToCoroutineDataList.forEach { (dispatcher, coroutineDataList) ->
+            coroutineDataList.forEach { data ->
+                if (data.descriptor.state == State.RUNNING) {
+                    dispatcherDropdownMenu.selectedItem = dispatcher
+                }
+            }
+        }
+
+        val selectedDispatcher = dispatcherDropdownMenu.selectedItem as String
+        val coroutineDataList = dispatcherToCoroutineDataList[selectedDispatcher]
+        if (!coroutineDataList.isNullOrEmpty()) {
+            updateCoroutineStackForest(coroutineDataList, suspendContextImpl)
+        }
+        panelContent.add(coroutineStacksWindowHeader)
+        panelContent.add(forest)
+        add(panelContent)
+    }
+
+    private fun updateCoroutineStackForest(
+        coroutineDataList: MutableList<CoroutineInfoData>,
+        suspendContextImpl: SuspendContextImpl
+    ) {
+        forest.removeAll()
+        val root = Node()
+        val coroutineStackForest = suspendContextImpl.buildCoroutineStackForest(
+            root,
+            coroutineDataList
+        )
+        forest.add(coroutineStackForest)
+        updateUI()
+    }
+
+    private fun SuspendContextImpl.buildCoroutineStackForest(
+        rootValue: Node,
+        coroutineDataList: MutableList<CoroutineInfoData>,
+    ): JBScrollPane {
+        val lastRunningStackFrame = buildStackFrameGraph(coroutineDataList, rootValue)
+        val coroutineTraces = createCoroutineTraces(rootValue)
+        return createCoroutineTraceForest(coroutineTraces, lastRunningStackFrame)
+    }
+
+    private fun SuspendContextImpl.createCoroutineTraceForest(traces: List<CoroutineTrace?>, lastRunningStackFrame: String): JBScrollPane {
+        val componentData: MutableList<Component> = mutableListOf()
+        var previousListSelection: JBList<*>? = null
+        val averagePreferredWidth = calculateAveragePreferredWidth(traces)
```

Comment:
Creating all stack traces just to calculate the average width is very expensive, let's calculate necessary components such as sum of widths and vertices number while we are creating vertices from coroutine traces